pkg/kind/kong_proxy_cluster_configuration.go

```diff
@@ -0,0 +1,98 @@
+package kind
+
+import (
+	"context"
+	"os"
+
+	"github.com/google/uuid"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/kong/kubernetes-testing-framework/pkg/metallb"
+)
+
+// -----------------------------------------------------------------------------
+// ClusterConfiguration - Public Types
+// -----------------------------------------------------------------------------
+
+const (
+	// EnvKeepCluster is the environment variable that can be set to "true" in order
+	// to circumvent teardown during cleanup of clusters in order to allow a user to inspect them instead.
+	EnvKeepCluster = "KIND_KEEP_CLUSTER"
+)
+
+// ClusterConfigurationWithKongProxy is an object representing a Kind cluster's configuration and can effectively
+// be used as a factory for kind cluster deployments. Clusters created from these configurations are
+// opinionated, and will always automatically pre-deploy a Kong proxy service.
+type ClusterConfigurationWithKongProxy struct {
+	// DockerNetwork indicates the name of the Docker network to use for LoadBalancer IPs
+	DockerNetwork string
+
+	// EnableMetalLB instructions the deployment of MetalLB to support provisioning LoadBalancer Services in the cluster.
+	EnableMetalLB bool
+}
+
+// -----------------------------------------------------------------------------
+// ClusterConfiguration - Factory Methods
+// -----------------------------------------------------------------------------
+
+// Deploy is a factory method to generate kind.Cluster objects given the configuration, with new names being selected on each deploy.
+func (c *ClusterConfigurationWithKongProxy) Deploy(ctx context.Context) (Cluster, chan ProxyReadinessEvent, error) {
+	name := uuid.New().String()
+
+	if c.DockerNetwork == "" {
+		c.DockerNetwork = DefaultKindDockerNetwork
+	}
+
+	err := CreateClusterWithKongProxy(name)
+	if err != nil {
+		return nil, nil, err
```

Comment:
nit, optional:

```suggestion
		return nil, nil, errors.Wrap(err, "CreateClusterWithKongProxy")
```

(requires `goimports`, and propagation to other `return err` places)

Reply:
Can you provide further context? I need some help understanding the motivation to add this :thinking:

Reply:
If a function call, nested 2 or 3 layers of function calls deep, returns an error, and its callers pass on the error verbatim, the result is that the high-level function `Deploy()` returns a very low-level and opaque error (say, "json marshal failed" or "command returned 1") without any context.

Wrapping passed errors like this helps attach some context to the error message as to where exactly the failure happened. I view this as an approximation of a call stack attached to an exception.

Reply:
Since error wrapping became native functionality with the `%w` formatting directive, I'm inclined to just pass the error up here and let the caller decide if they want to wrap it for additional context. Do you think that's acceptable, or does using the 3rd party wrapping library still seem preferable to you?

Reply:
```suggestion
		return nil, nil, fmt.Errorf("CreateClusterWithKongProxy failed: %w", err)
```